packages/vite/src/node/plugins/asset.ts

```diff
@@ -249,19 +250,41 @@ export async function urlToBuiltUrl(
   url: string,
   importer: string,
   config: ResolvedConfig,
-  pluginContext: PluginContext
+  ctx: PluginContext
 ): Promise<string> {
+  const builtUrl = await resolveBuiltUrl(url, config, ctx)
+  if (builtUrl) {
+    return builtUrl
+  }
+
   if (checkPublicFile(url, config)) {
     return config.base + url.slice(1)
   }
+
   const file = url.startsWith('/')
     ? path.join(config.root, url)
     : path.join(path.dirname(importer), url)
+
   return fileToBuiltUrl(
     file,
     config,
-    pluginContext,
+    ctx,
     // skip public check since we just did it above
     true
   )
 }
+
+async function resolveBuiltUrl(
+  url: string,
+  config: ResolvedConfig,
+  ctx: PluginContext
+) {
+  for (const { resolveBuiltUrl } of config.plugins) {
+    if (resolveBuiltUrl) {
+      const result = await resolveBuiltUrl.call(ctx, url)
+      if (result) {
+        return result
```

Comment:
**thought (non-blocking, dx):** IMO always returning at least something increases the developer experience
So you could extract a `let result` out of the for-loop and then when the `if` was triggered call a `break`
At the bottom of the function body you can then return the `result` (that could be `undefined`)